### ISSUE TITLE
Zipped streams of Free blow up with CCE

### ISSUE THREAD

**Issue by tpolecat** [functional-streams-for-scala/fs2#1099]
Running a zipped `Stream[Free` translated to `IO` blows up with a `ClassCastException` as was seen some time ago. This is with fs2 0.10.2

```scala
import cats._
import cats.effect.IO
import cats.free.Free
import cats.implicits._
import fs2.Stream

object FreeGoBoom {

  val trans: Free[Id, ?] ~> IO =
    Free.foldMap(λ[Id ~> IO](_.pure[IO]))

  val empty: Stream[Free[Id, ?], Nothing] =
    Stream.empty.covary[Free[Id, ?]]

  def main(args: Array[String]): Unit =
    (empty zip empty)
      .translate(trans)
      .compile
      .drain
      .unsafeRunSync

}
```

The exception looks like

```
[error] (run-main-3f) java.lang.ClassCastException: cats.effect.IO$Bind cannot be cast to cats.free.Free
[error] java.lang.ClassCastException: cats.effect.IO$Bind cannot be cast to cats.free.Free
[error] 	at FreeGoBoom$$anon$1.apply(SmartGcalImporter.scala:13)
[error] 	at fs2.internal.Algebra$AlgEffectSyntax$.translate$extension(Algebra.scala:54)
[error] 	at fs2.internal.Algebra$.translate0(Algebra.scala:564)
...
```

I'm sorry.

**Comment by AdamChlupacek** [comment]
The issue here is that in zip we are running the two stream from within the stepLeg with an `F (Sync[F])` that is taken from the scope. As such suddenly the zip of the two `Free` have an `F` that is an `IO` thus this happens. 

One solution for this is that the zip will be taking the `Sync` for its current `F` thus getting the proper values.

**Comment by AdamChlupacek** [comment]
This has an obvious additional requirement for `Sync[F]` when zipping/merging deterministically two stream. Not sure if this is acceptable for us to require in case of pure streams and such.

**Comment by mpilquist** [comment]
@AdamChlupacek I really don't want to require a `Sync[F]` constraint on `zip` and derivatives. I think we should be able to zip pure streams.

**Comment by AdamChlupacek** [comment]
@mpilquist I agree, thats why I did not really made a PR for it. But as of now I am not entirely sure how to do this proper, we will probably need a new case class in Algebra. I may get something like that together today. It may not look too bad.

**Comment by pchlupacek** [comment]
@mpilquist @AdamChlupacek I think this comes to root of the design now, and essentially the issue is that `Scope` now requires `F`. Albeit we could write Scope[F].translate(f: F ~> G): Scope[G], the issue is that we have essentially F~>G available in translate but we need to do Scope[G] ~> Scope[F], so the reverse way. 

The trick with re-using the Sync[F] from scope in implementation of `zip` is obviously conflicting with `translate`. 

We may have mulitple options: 
- Reintroduce  `Tee` or similar to do it w/o `sync` requirement.  
- Add `StepLeg` directly in Algebra. Chances are that it will. Not sure though if that won't clutter code with another `translate`-like pattern match. 

Let me know if you have other ideas.

**Comment by tpolecat** [comment]
So, after this patch our code runs a bit longer before blowing up with a similar error:

```
Importing Flamingos2 ...
Importing GMOS-N ...
[error] (run-main-1) java.lang.ClassCastException: cats.effect.IO$Bind cannot be cast to cats.free.Free
[error] java.lang.ClassCastException: cats.effect.IO$Bind cannot be cast to cats.free.Free
[error]     at doobie.util.transactor$Transactor$$anon$6$$anon$8.apply(transactor.scala:184)
[error]     at fs2.internal.Algebra$AlgEffectSyntax$.translate$extension(Algebra.scala:57)
[error]     at fs2.internal.Algebra$.fs2$internal$Algebra$$translate0(Algebra.scala:614)
```

The empty stream example above does work now, but in our real codebase with nontrivial/non-empty `Stream[Free` it's still failing. 

If we remove the `zip` call our code runs fine on 0.10.2, and we have a patch in place so we're no longer blocked.

I can re-minimize if you need another example but it will take a while because I expect I'll have to traverse into doobie. Let me know if you need it.

Thanks!

**Comment by AdamChlupacek** [comment]
@tpolecat Would be great to have another minimised example. I will give a bit more time to go through source, but minimised example/pointer to place in doobie source where the error is occurring would be appreciated.

**Comment by tpolecat** [comment]
Ok, will do. 👍

**Comment by tpolecat** [comment]
Funny, after starting over from our application code I arrived at something almost as simple.

```scala
import cats._
import cats.free._
import cats.effect.IO
import cats.implicits._
import fs2.Stream

object SmartGcalImporter extends DoobieClient {

  val trans: Free[Id, ?] ~> IO =
    Free.foldMap(λ[Id ~> IO](_.pure[IO]))

  val one: Stream[Free[Id, ?], Int] =
    Stream.eval(Free.liftF(1 : Id[Int]))

  def main(args: Array[String]): Unit =
    ((one ++ one) zip one)
      .translate(trans)
      .compile
      .drain
      .unsafeRunSync

}
```

**Comment by AdamChlupacek** [comment]
Ok, this is actually my bad. The translate of the `RunLeg(_)` is still unsound. At the moment it actually gets translated multiple times ie `Free -> IO -> IO`, thus this obvious error.

Now to think of a solution...

**Comment by AdamChlupacek** [comment]
Good news, I have a working fix, bad news it is quite ugly atm :/ will try to clean in up a little bit.

Let me as well write few more tests for this, so that we are sure that translate will really work now.

**Comment by tpolecat** [comment]
👍 Thanks for your quick attention!

**Comment by mpilquist** [comment]
Fixed by #1102
